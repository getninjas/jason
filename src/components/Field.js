import React, { Component, cloneElement } from 'react';
import PropTypes from 'prop-types';

const propTypes = {
  label: PropTypes.string.isRequired,
  id: PropTypes.string.isRequired,
  children: PropTypes.element.isRequired,
  errorMessage: PropTypes.string,
  value: PropTypes.string,
  isFluid: PropTypes.Boolean,
};

const defaultProps = {
  errorMessage: '',
  value: '',
  isFluid: true,
};

export default class Field extends Component {
  constructor(props) {
    super(props);

    this.wrapperStyle = this.props.isFluid ? 'form__field form__field--fluid input' : 'form__field input';
    this.spanStyle = 'form__message form__message--invalid space-element-small error';
  }

  get style() {
    if (this.props.errorMessage) {
      return 'form__input form__input--invalid';
    }

    return 'form__input';
  }

  render() {
    return (
      <div className={this.wrapperStyle}>
        <label htmlFor={this.props.id} className="form__label">
          { this.props.label }
        </label>

        { cloneElement(this.props.children, { style: this.style }) }

        <span className={this.spanStyle}>
          { this.props.errorMessage }
        </span>
      </div>
    );
  }
}

Field.propTypes = propTypes;
Field.defaultProps = defaultProps;
