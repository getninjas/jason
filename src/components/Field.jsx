import React from 'react';
import PropTypes from 'prop-types';

export default class Field extends React.Component {
  render() {
    return (
      <div className="form__field form__field--fluid input">
        <label htmlFor={this.props.id} className="form__label">
          { this.props.label }
        </label>

        { this.props.children }

        <span className="form__message form__message--invalid space-element-small error">
          { this.props.errorMessage }
        </span>
      </div>
    );
  }
}

Field.defaultProps = {
  label: '',
  id: '',
  errorMessage: ''
}

Field.propTypes = {
  label: PropTypes.string.isRequired,
  id: PropTypes.string.isRequired,
  children: PropTypes.any.isRequired,
  errorMessage: PropTypes.string,
}
