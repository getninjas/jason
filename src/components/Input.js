import React, { Component, createRef } from 'react';
import PropTypes from 'prop-types';
import IMask from 'imask';
import { maxLengthTrim, getInputType } from '../helpers/input';

const propTypes = {
  id: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  onFieldChange: PropTypes.func.isRequired,
  onFieldBlur: PropTypes.func.isRequired,
  title: PropTypes.string,
  placeholder: PropTypes.string,
  type: PropTypes.string,
  required: PropTypes.bool,
  initialValue: PropTypes.any,
  values: PropTypes.array,
  style: PropTypes.string,
  minLength: PropTypes.number,
  maxLength: PropTypes.number,
};

const defaultProps = {
  placeholder: '',
  required: false,
  title: '',
  type: 'text',
  initialValue: '',
  values: [],
  style: 'form__input',
  minLength: 3,
  maxLength: 255,
};

export default class Input extends Component {
  constructor(props) {
    super(props);

    this.state = {
      values: this.props.values,
      value: this.props.initialValue ? this.props.initialValue : '',
    };

    this.ref = createRef();
    this.onChange = this.onChange.bind(this);
    this.onChangeCheck = this.onChangeCheck.bind(this);
    this.onBlur = this.onBlur.bind(this);
    this.mask = null;
  }

  onChange(evt) {
    const inputValue = maxLengthTrim(evt.target.value, this.props.maxLength);

    this.props.onFieldChange({
      value: inputValue,
      id: this.props.id,
    });

    this.setState({ value: inputValue });
  }

  onChangeCheck(evt) {
    console.log('evt', evt);
    let idList = [];
    if (!this.state.value.length) {
      idList.push(parseInt(evt.target.id, 10));
    } else {
      idList = this.state.value.filter(valor => valor !== parseInt(evt.target.id, 10));

      if (idList.length === this.state.value.length) {
        idList.push(parseInt(evt.target.id, 10));
      }
    }

    this.props.onFieldChange({
      id: this.props.id,
      value: idList,
    });

    this.setState({ value: idList });
  }

  onBlur(evt) {
    this.props.onFieldBlur({
      value: evt.target.value,
      id: this.props.id,
      required: this.props.required,
      type: this.props.type,
      minLength: this.props.minLength,
    });
  }

  componentDidMount() {
    const { type } = this.props;

    if (type === 'phone') {
      this.mask = new IMask(this.ref.current, { mask: '(00) 00000-0000' });
      this.mask.on('complete', () => {
        this.props.onFieldChange({
          value: this.mask.value,
          id: this.props.id,
        });

        this.setState({ value: this.mask.value });
      });
    }
  }

  render() {
    const {
      type,
      id,
      name,
      title,
      style,
      placeholder,
      required,
      minLength,
      maxLength,
    } = this.props;

    const prepareComponent = () => {
      if (type === 'checkbox') {
        return (
          <div required={required ? 'true' : 'false'}>
            {this.state.values.map((elem, idx) => (
                <label key={`${elem.databaseId}-${idx}`} htmlFor={elem.databaseId}>
                  <input
                    type={getInputType(type)}
                    id={elem.databaseId}
                    name={name}
                    title={title}
                    className={style}
                    value={elem.value}
                    onChange={this.onChangeCheck}
                    ref={this.ref} />
                    {elem.value === 'OTHER' ? (
                      <input type="text" />
                    ) : elem.value}
                </label>
              ))}
          </div>
        );
      }

      return (
        <input
          type={getInputType(type)}
          id={id}
          name={name}
          title={title}
          className={style}
          placeholder={placeholder}
          required={required ? 'true' : 'false'}
          value={this.state.value}
          onChange={this.onChange}
          onBlur={this.onBlur}
          minLength={minLength}
          maxLength={maxLength}
          ref={this.ref} />
      );
    };

    return prepareComponent();
  }
}

Input.propTypes = propTypes;
Input.defaultProps = defaultProps;
