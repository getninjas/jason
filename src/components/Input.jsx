import React from 'react';
import PropTypes from 'prop-types';

export default class Input extends React.Component {
  constructor() {
    super();

    this.state = {
      value: '',
    }

    this.onChange = this.onChange.bind(this);
  }

  onChange(evt) {
    this.setState({ value: evt.target.value });
  }

  componentWillMount() {
    this.setState({ value: this.props.value });
  }

  render() {
    return (
      <input
        type={this.props.type==='phone'? 'tel': this.props.type}
        id={this.props.id}
        name={this.props.name}
        title={this.props.title}
        className="form__input"
        placeholder={this.props.placeholder}
        required={this.props.required ? 'true' : 'false'}
        value={this.state.value}
        onChange={this.onChange} />
    );
  }
}

Input.defaultProps = {
  id: '',
  placeholder: '',
  required: false,
  name: '',
  type: 'text',
  title: '',
  value: '',
}

Input.propTypes = {
  id: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired,
  placeholder: PropTypes.string,
  type: PropTypes.string,
  required: PropTypes.bool,
  value: PropTypes.any,
}
