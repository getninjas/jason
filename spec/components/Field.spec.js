import React from 'react';
import renderer from 'react-test-renderer';
import Field from '../../src/components/Field';
import Input from '../../src/components/Input';
import { enzymeConfig, shallow } from '../enzymeConfig';

enzymeConfig();

describe('Field', () => {
  it('renders defaultProps', () => {
    const component = renderer.create(
      <Field label={'test'} id={'test'} errorMessage={'Erro!'} value={'test value'} isFluid={true}>
        <Input id={'bora-pra-action'} name={'nameTest'} onFieldChange={() => {}} />
      </Field>,
    );

    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('label htmlFor matches id prop', () => {
    const component = shallow(
      <Field label={'test'} id={'bora-pra-action'}>
        <Input id={'bora-pra-action'} name={'nameTest'} onFieldChange={() => {}} />
      </Field>,
    );

    expect(component.find('label').prop('htmlFor')).toEqual('bora-pra-action');
  });

  describe('.style', () => {
    describe('with erroMessage', () => {
      it('returns string containing --invalid', () => {
        const component = shallow(
          <Field label={'test'} id={'bora-pra-action'} errorMessage={'Erro!'}>
            <Input id={'bora-pra-action'} name={'nameTest'} onFieldChange={() => {}} />
          </Field>,
        );

        expect(component.instance().style).toMatch(/--invalid/);
      });
    });

    describe('without erroMessage', () => {
      it('returns string not containing --invalid', () => {
        const component = shallow(
          <Field label={'test'} id={'bora-pra-action'}>
            <Input id={'bora-pra-action'} name={'nameTest'} onFieldChange={() => {}} />
          </Field>,
        );

        expect(component.instance().style).not.toMatch(/--invalid/);
      });
    });

    describe('with isFluid', () => {
      it('returns className contains form__field--fluid', () => {
        const component = shallow(
          <Field label={'test'} id={'bora-pra-action'} isFluid={true}>
            <Input id={'bora-pra-action'} name={'nameTest'} onFieldChange={() => { }} />
          </Field>,
        );

        expect(component.instance().wrapperStyle).toMatch(/form__field--fluid/);
      });
    });

    describe('without isFluid', () => {
      it('returns className not contains form__field--fluid', () => {
        const component = shallow(
          <Field label={'test'} id={'bora-pra-action'} isFluid={false}>
            <Input id={'bora-pra-action'} name={'nameTest'} onFieldChange={() => { }} />
          </Field>,
        );

        expect(component.instance().wrapperStyle).not.toMatch(/form__field--fluid/);
      });
    });
  });
});
