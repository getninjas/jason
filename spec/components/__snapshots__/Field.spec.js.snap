// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Field renders defaultProps 1`] = `
<div
  className="form__field"
>
  <label
    className="form__label"
    htmlFor="test"
  >
    test
  </label>
  <input
    className="form__input form__input--invalid"
    id="bora-pra-action"
    maxLength={255}
    minLength={1}
    name="nameTest"
    onBlur={[Function]}
    onChange={[Function]}
    placeholder=""
    required="false"
    title=""
    type="text"
    value=""
  />
  <span
    className="form__message form__message--invalid space-element-small error"
  >
    Erro!
  </span>
</div>
`;
